### ISSUE TITLE
Missed optimization with `__attribute__((pure))`

### ISSUE THREAD

**Issue by OfekShilon** [llvm/llvm-project#53102]
```c++
int somefunc(const int&) __attribute__((pure));
void nothing();

int f(int i) {
    i = somefunc(i);
    i++;
    nothing();
    i++;
    nothing();
    i++;
    return i;
}
```
The emitted code loads, increments, and stores `i` 3 times : https://godbolt.org/z/7YW7187s1
In gcc it doesn't: https://godbolt.org/z/Ye9aYn4MG
Also, if `somefunc` is made to return void, the optimization does kick in: https://godbolt.org/z/hsEdPThsf

**Comment by zero9178** [comment]
It looks to me like LLVM thinks that `somefunc` might capture the pointer (aka store it eg. in a global variable), hence `nothing` could modify it which would require a reload. 

Clearly this is a wrong assumption though as it is marked `pure` which gets translated to LLVMs `readonly` which in the language reference also mentions `If a readonly function writes memory visible to the program, or has other side-effects, the behavior is undefined. If a function writes to a readonly pointer argument, the behavior is undefined.`. 

Your second example where it returns `void` further shows this. LLVM sees a `readonly` function that does not return anything and therefore simply removes the call. Because of that, the capture that LLVM previously thought might happen in the previous example is now gone and it manages to remove the loads.

**Comment by OfekShilon** [comment]
I'm inspecting my codebase with a modified version of LLVM's [optviewer](https://github.com/llvm/llvm-project/tree/main/llvm/tools/opt-viewer), and come across tons of 'load clobbered by call' optimization failures. In fact, these are most the failed optimizations I see. I'm hoping `__attribute__((pure))` and `__attribute__((const))` can aid in eliminating some of those.

**Comment by OfekShilon** [comment]
Need to add: gcc *does* use this attribute to optimize in this exact case:
https://godbolt.org/z/W566WaE7q

**Comment by jdoerfert** [comment]
> It looks to me like LLVM thinks that `somefunc` might capture the pointer (aka store it eg. in a global variable), hence `nothing` could modify it which would require a reload.

Might capture, yes. Store it, no. The issue is returning it: https://godbolt.org/z/75dna7K1a

> Clearly this is a wrong assumption though as it is marked `pure` which gets translated to LLVMs `readonly` which in the language reference also mentions `If a readonly function writes memory visible to the program, or has other side-effects, the behavior is undefined. If a function writes to a readonly pointer argument, the behavior is undefined.`.

No to the wrong assumption part, see example above. Returning any bit of `&i` captures `i`, assuming we do not throw a lot of analysis on it.

> Your second example where it returns `void` further shows this. LLVM sees a `readonly` function that does not return anything and therefore simply removes the call. Because of that, the capture that LLVM previously thought might happen in the previous example is now gone and it manages to remove the loads.

Right.



> I'm inspecting my codebase with a modified version of LLVM's [optviewer](https://github.com/llvm/llvm-project/tree/main/llvm/tools/opt-viewer), and come across tons of 'load clobbered by call' optimization failures. In fact, these are most the failed optimizations I see. I'm hoping `__attribute__((pure))` and `__attribute__((const))` can aid in eliminating some of those.

`noescape` (in C/C++) or `nocapture` (in IR) is what you are looking for: https://godbolt.org/z/7EY9xKeoe

**Comment by OfekShilon** [comment]
@jdoerfert :
> Might capture, yes. Store it, no. The issue is returning it: https://godbolt.org/z/75dna7K1a

This godbolt is on gcc. If clang 14 is used, the `somefunc`/`nothing` calls are eliminated altogether when the definitions are available: https://godbolt.org/z/7abvxzdxT
When the definitions aren't available, in this variant `__attribute__((pure))` does make optimization kick in, even without a return value from `somefunc`: https://godbolt.org/z/3qv4jf1bG    ( turn pure on/off to see)
Am I missing something?

 > `noescape` (in C/C++) or `nocapture` (in IR) is what you are looking for: https://godbolt.org/z/7EY9xKeoe

Thanks! Will definitely use in the future.

**Comment by jdoerfert** [comment]
> Am I missing something?

I should have been more explicit. I used GCC to show the extra logic it puts in place to determine if the value didn't unexpectedly change (due to the returned integer potentially capturing the value). At least that is what I think it's doing :D

**Comment by OfekShilon** [comment]
@jdoerfert   Not sure I follow.  Are you saying that gcc is wrong in coalescing the 3 increments?
What do you mean by 'capture a pointer through a return value'?  I can come up with a scenario where the return type has a non-default ctor that does crazy stuff like store argument data somewhere, but that clearly violates the readonly attribute.
Can you please give a concrete example of what you think the compiler is afraid of?   For example - an implementation of `somefunc` and `nothing` in the example that (1) doesn't violate the `pure` attribute, (2) makes coalescing the 3 increments illegal?
I mean, how can merely returning a value make the argument escape?

**Comment by jdoerfert** [comment]
> @jdoerfert Not sure I follow. Are you saying that gcc is wrong in coalescing the 3 increments?

No. It is adding sufficient logic to verify that coalescing is valid, at least that is what it looks like to me.

> What do you mean by 'capture a pointer through a return value'? I can come up with a scenario where the return type has a non-default ctor that does crazy stuff like store argument data somewhere, but that clearly violates the readonly attribute. Can you please give a concrete example of what you think the compiler is afraid of? For example - an implementation of `somefunc` and `nothing` in the example that (1) doesn't violate the `pure` attribute, (2) makes coalescing the 3 increments illegal? I mean, how can merely returning a value make the argument escape?

My first comment has an example of what the compiler is afraid of: https://godbolt.org/z/75dna7K1a

**Comment by OfekShilon** [comment]
> > @jdoerfert Not sure I follow. Are you saying that gcc is wrong in coalescing the 3 increments?
> 
> No. It is adding sufficient logic to verify that coalescing is valid, at least that is what it looks like to me.

I was referring to my very first example. gcc is does coalesce the 3 increments, even with a return value from somefunc: https://godbolt.org/z/Ye9aYn4MG. Are you saying this is a mistake?

> My first comment has an example of what the compiler is afraid of: https://godbolt.org/z/75dna7K1a

If I'm guessing correctly what you mean, you say that gcc inserts explicit code to check whether `i == 7 + 11` and so there must be _some_ "capture" (?) it is afraid of.  (1) That doesn't explain what gcc was afraid of, just claims that it must be afraid of _something_.  (2) In my first example (https://godbolt.org/z/Ye9aYn4MG) the same gcc deduces from `__attribute__((pure))` that `somefunc` doesn't capture anything, and so feels free to coalesce the `i` increments. 

So we probably shouldn't take gcc as some ground truth, and try to explicitly understand whether `__attribute__((pure))` leaves some capture scenario to be afraid of.  I think there isn't one - did I miss something? Do you have such a scenario?

**Comment by OfekShilon** [comment]
@jdoerfert here's another example. If you let `somefunc` return a value but change its type from `int` to `int&` - the coalescing optimization does kick in: https://godbolt.org/z/8c5fjjf95
So I don't see how having a return value can justify canceling the optimization (or that there is any capture by a return value).  I believe it is just a missed llvm optimization.

**Comment by jdoerfert** [comment]
> > 
> 
> I was referring to my very first example. gcc does coalesce the 3 increments, even with a return value from somefunc: https://godbolt.org/z/Ye9aYn4MG. Are you saying this is a mistake?

No. I never said there is a mistake. The example is sufficiently easy. Even if the pointer escape via the return one can (and GCC does) determine the return value is not used as a pointer.


> > My first comment has an example of what the compiler is afraid of: https://godbolt.org/z/75dna7K1a
> 
> If I'm guessing correctly what you mean, you say that gcc inserts explicit code to check whether `i == 7 + 11` and so there must be _some_ "capture" (?) it is afraid of. (1) That doesn't explain what gcc was afraid of, just claims that it must be afraid of _something_. (2) In my first example (https://godbolt.org/z/Ye9aYn4MG) the same gcc deduces from `__attribute__((pure))` that `somefunc` doesn't capture anything - even with a return value - and so feels free to coalesce the `i` increments.
> 
> So we probably shouldn't take gcc as some ground truth, and try to explicitly understand whether `__attribute__((pure))` leaves some capture scenario to be afraid of. I think there isn't one - did I miss something? Do you have such a scenario?

I literally showed you an example in which the pure function captures the pointer. GCC realizes that I was potentially doing that and added a check. LLVM cannot add a check and bails. My example is what you are looking for. It has nothing to do with GCC, it's just an example (one of the easiest for that matter) in which we use a pure function to escape a pointer. I mean, returning the pointer does arguably leak the bits of the pointer, agreed?


> @jdoerfert here's another example. If you let `somefunc` return a value but change its type from `int` to `int&` - the coalescing optimization does kick in: https://godbolt.org/z/8c5fjjf95 

I don't understand what this has to do with anything. Returning int& allows to escape  the argument pointer just as well. 

> So I don't see how having a return value can justify canceling the optimization (or that there is any capture by a return value). I believe it is just a missed llvm optimization.

Let me try one more time: A return values (even bool) allow you to escape the bits of the pointer. Once potentially escaped, compiles have 3 choices:
1) Find out if the returned bits might be used to access the pointer. If not, you're fine. <- This is your initial example. GCC notices they cannot be used as pointer and thereby ignores capturing.
2) If the returned bits might be used to access the pointer, insert a check to verify if they have been accessed and changed. <- My GCC example.
3) Bail out. <- LLVM for all examples.

**Comment by jdoerfert** [comment]
> In practical terms, what we're missing here is "If the call is both readonly, nounwind, willreturn and the return value is not used, then all pointer operands can be marked nocapture", right?

Technically, this is an option but the non-helpful one. Readonly + nounwind + willreturn and unused return value -> the call is dead.

What you need is a scoped check for the return value to be captured (or used as a pointer).
Scoped because any use I can come up with is either effectively dead or leaves the caller one way or another (if only as control information). That means you can only argue about function local things, namely allocas, which is what is passed in the original example.

**Comment by nikic** [comment]
> > In practical terms, what we're missing here is "If the call is both readonly, nounwind, willreturn and the return value is not used, then all pointer operands can be marked nocapture", right?
> 
> Technically, this is an option but the non-helpful one. Readonly + nounwind + willreturn and unused return value -> the call is dead.

Uh yeah, never mind me, that suggestion didn't make any sense :)

**Comment by OfekShilon** [comment]
> 
> > @jdoerfert here's another example. If you let `somefunc` return a value but change its type from `int` to `int&` - the coalescing optimization does kick in: https://godbolt.org/z/8c5fjjf95
> 
> I don't understand what this has to do with anything. Returning int& allows to escape the argument pointer just as well.
> 

In theory your explanation would hold in this case (returning `int&`) as well, and yet it seems it doesn't - the coalescing optimization does kick in. Why?

> Let me try one more time: A return values (even bool) allow you to escape the bits of the pointer. Once potentially escaped, compiles have 3 choices:
> 
> 1. Find out if the returned bits might be used to access the pointer. If not, you're fine. <- This is your initial example. GCC notices they cannot be used as pointer and thereby ignores capturing.
> 2. If the returned bits might be used to access the pointer, insert a check to verify if they have been accessed and changed. <- My GCC example.
> 3. Bail out. <- LLVM for all examples.

Thanks for your effort to explain.  I (think I) understand this reasoning in _your_ example code, but mine was rather far from it: `somefunc`'s return value was never used passed outside the main function. 
So IIUC you're essentially saying this bail-out is the root cause of this missed optimization? LLVM just doesn't implement the needed extra analysis?

**Comment by jdoerfert** [comment]
> > > @jdoerfert here's another example. If you let `somefunc` return a value but change its type from `int` to `int&` - the coalescing optimization does kick in: https://godbolt.org/z/8c5fjjf95
> > 
> > 
> > I don't understand what this has to do with anything. Returning int& allows to escape the argument pointer just as well.
> 
> In theory your explanation would hold in this case (returning `int&`) as well, and yet it seems it doesn't - the coalescing optimization does kick in. Why?

GCC always applies coalescing, it chooses 1. from below. Why is this different from the initial example?

> > Let me try one more time: A return values (even bool) allow you to escape the bits of the pointer. Once potentially escaped, compiles have 3 choices:
> > 
> > 1. Find out if the returned bits might be used to access the pointer. If not, you're fine. <- This is your initial example. GCC notices they cannot be used as pointer and thereby ignores capturing.
> > 2. If the returned bits might be used to access the pointer, insert a check to verify if they have been accessed and changed. <- My GCC example.
> > 3. Bail out. <- LLVM for all examples.
> 
> Thanks for your effort to explain. I (think I) understand this reasoning in _your_ example code, but mine was rather far from it: `somefunc`'s return value was never used passed outside the main function. So IIUC you're essentially saying this bail-out is the root cause of this missed optimization? LLVM just doesn't implement the needed extra analysis?

Yes. What we need is described here: https://github.com/llvm/llvm-project/issues/53102#issuecomment-1088907652

**Comment by OfekShilon** [comment]
> > > > @jdoerfert here's another example. If you let `somefunc` return a value but change its type from `int` to `int&` - the coalescing optimization does kick in: https://godbolt.org/z/8c5fjjf95
> > > 
> > > 
> > > I don't understand what this has to do with anything. Returning int& allows to escape the argument pointer just as well.
> > 
> > 
> > In theory your explanation would hold in this case (returning `int&`) as well, and yet it seems it doesn't - the coalescing optimization does kick in. Why?
> 
> GCC always applies coalescing, it chooses 1. from below. Why is this different from the initial example?
> 

Sorry - I meant to send it with *clang*:  https://godbolt.org/z/xesE5o14x
clang does optimize when the return type is `int&`.

**Comment by jdoerfert** [comment]
> > > > > @jdoerfert here's another example. If you let `somefunc` return a value but change its type from `int` to `int&` - the coalescing optimization does kick in: https://godbolt.org/z/8c5fjjf95
> > > > 
> > > > 
> > > > I don't understand what this has to do with anything. Returning int& allows to escape the argument pointer just as well.
> > > 
> > > 
> > > In theory your explanation would hold in this case (returning `int&`) as well, and yet it seems it doesn't - the coalescing optimization does kick in. Why?
> > 
> > 
> > GCC always applies coalescing, it chooses 1. from below. Why is this different from the initial example?
> 
> Sorry - I meant to send it with _clang_: https://godbolt.org/z/xesE5o14x clang does optimize when the return type is `int&`.

Oh... that's interesting. I was looking (and talking about) the GCC output. 

So I tried to figure out where the difference comes from and I'm not sure right now.
I'll try to revisit this at some point soon. Now I'm curious why we perform this transformation.

**Comment by OfekShilon** [comment]
@jdoerfert  I'm curious if anything came up

**Comment by jdoerfert** [comment]
> @jdoerfert I'm curious if anything came up

I did not manage to spend any more time on this, sorry. I'll keep it in mind but it might be worth to ask in the LLVM forum/list too. Might get someone else interested enough too investigate.
